File: algorithms/sorting/merge_sort/merge_sort.cpp
#include <iostream>

using namespace std;

#define REP(i, n) for(int i = 0; i < n; ++i)

void merge(int *a, int p, int q, int r)
{
  int n1 = (q - p + 1), n2 = (r - q);

  int L[n1], M[n2];

  REP(i, n1) L[i] = a[p + i];
  REP(j, n2) M[j] = a[q + j + 1];

  int i = 0, j = 0;

  for(int k = p; k <= r; k++) {
    if ((i < n1) && ((j >= n2) || (L[i] <=  M[j]))) {
      a[k] = L[i];
      ++i;
    }
    else {
      a[k] = M[j];
      j++;
    }
  }
}

void merge_sort(int *a, int p, int r)
{
  if (p < r) {
    int q = (p + r)/2 ;
    merge_sort(a, p, q);
    merge_sort(a, q + 1, r);
    merge(a, p, q, r);
  }
}

void display_array(int *a, int size)
{
  cout << endl;
  REP(i, size) cout << a[i] << "  ";
  cout << endl;
}

int main()
{
  int *a;
  int size;

  cout << "What size of array do you want? ";
  cin >> size;

  a = new int[size];
  cout << "Enter elements of the array : ";
  REP(i, size) cin >> a[i];

  cout << "Current array : " << endl;
  display_array(a, size);
  
  merge_sort(a, 0, size - 1);

  cout << "Sorted array : " << endl;
  display_array(a, size);
}

File: algorithms/sorting/merge_sort/merge_sort.rb
class Array
  def merge_sort!(x = 0, z = (size - 1))
    if x < z
      y = (x + z)/2
      merge_sort!(x, y)
      merge_sort!(y + 1, z)
      array_merge(x, y, z)
    end
  end

  def array_merge(x, y, z)
    n1 = y - x + 1
    n2 = z - y

    l = self[x...(x + n1)]
    m = self[(y + 1)...(y + 1 + n2)]

    i, j = [0, 0]
    (x..z).each do |k|
      if (i < n1) && ((j >= n2) || (l[i] <= m[j]))
        self[k] = l[i]
        i += 1
      else
        self[k] = m[j]
        j += 1
      end
    end
  end
end

def interact
  puts "What size array do you want? "
  size = gets.chomp.to_i

  array = []
  puts "Enter the array : "
  size.times { array.push(gets.chomp.to_i) }

  puts "Current array : "
  puts array.inspect

  array.merge_sort!

  puts "Sorted array : "
  puts array.inspect
end

interact
